### PR TITLE
Fix Unicode and emoji handling in JNI string conversions

### PR DESCRIPTION
**Pull request by 985892345**
### Background

Kotlin/JVM represents `String` values as UTF-16 `jstring` objects across JNI.

The existing bridge uses `GetStringUTFChars` and `NewStringUTF` for string conversion. However, these JNI APIs use Modified UTF-8, while the QuickJS C API expects standard UTF-8.

The two encodings behave similarly for ASCII and most BMP characters, but encode supplementary characters such as emoji differently:

- Standard UTF-8 encodes an emoji as a single four-byte sequence.
- Modified UTF-8 encodes its UTF-16 surrogate pair as two three-byte sequences.

Passing Modified UTF-8 through the QuickJS C API can therefore corrupt emoji and other supplementary Unicode characters.

### Changes

```text
Kotlin/JVM String
   UTF-16 jstring
         ⇅
 Shared JNI converter
         ⇅
UTF-8/WTF-8 bytes + explicit length
         ⇅
    QuickJS C API
         ⇅
JavaScript String
   UTF-16 code-unit semantics
```

- Valid Unicode content uses standard UTF-8, while unpaired UTF-16 surrogates are preserved internally with WTF-8.
- Evaluation, bindings, value mapping, exceptions, and module loading now use the shared converter.
- The CMake configuration now ensures that new JNI source files are included for every Android ABI.

### Compatibility

- No changes to the public Kotlin API, ABI, or existing call sites.
- No changes to the QuickJS string implementation.
- Existing behavior for ASCII, Chinese text, and other supported BMP characters remains unchanged.
- Emoji and other supplementary characters now use standard UTF-8 at the QuickJS C API boundary.
- WTF-8 is used only inside the JNI bridge and is never exposed as an external encoding.

### Tests

Added coverage for:

- Chinese text, combining characters, emoji, and supplementary-plane identifiers.
- JavaScript source, evaluation results, and bytecode.
- Binding names, arguments, return values, and object keys.
- Kotlin and JavaScript exception messages.
- ES module names, source, and compilation callbacks.
- Unpaired high and low UTF-16 surrogates and other string boundary cases.

### Verification

- Relevant automated tests and builds pass.
- Manually verified on macOS and Android that Unicode and emoji content is passed through and displayed correctly.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Unicode handling across JavaScript evaluation, bindings, objects, modules, errors, and bytecode.
  * Preserved embedded NUL characters and unpaired surrogates during JavaScript–Java string conversion.
  * Improved malformed-text and conversion-failure handling with clearer error behavior.
  * Module evaluation now respects full source length and rejects embedded NULs in module names.
  * Native builds now detect added or removed source files automatically.

* **Tests**
  * Added coverage for Unicode, emoji, supplementary characters, embedded NULs, and unpaired surrogates across bridge functionality.

### CODE REVIEW

File: quickjs/native/jni/exception_util.c [hunk 1]
```diff
@@ -66,7 +75,19 @@ void jni_throw_qjs_exception(JNIEnv *env, const char *format, ...) {
...
+                env,
+                cls_quick_js_exception(env),
+                method_quick_js_exception_init(env),
+                message);
+        if (exception != NULL) {
+            (*env)->Throw(env, exception);
+            (*env)->DeleteLocalRef(env, exception);
+        }
+        (*env)->DeleteLocalRef(env, message);
+    }
```
[the remark this answers is not in the document]
Comment (985892345) [reply quickjs/native/jni/exception_util.c]: Fixed in de43fe6. The formatting buffer length and allocation are now validated before writing, and the explicit-length throw path falls back to a QuickJsException only when JNI has no exception pending. The full JVM suite and Android Debug/Release builds pass.

**Comment by 985892345** [comment]
Addressed the remaining CodeRabbit outside-diff finding in de43fe6: JavaScript error names and messages now use JS_ToCStringLen, and explicit byte lengths are preserved through both typed-error and fallback QuickJsException construction. The same length-aware composition is also used for thrown JavaScript errors. Added regression coverage for returned and thrown Error values containing embedded NUL characters; the full JVM suite and Android Debug/Release builds pass.

File: quickjs/native/jni/js_value_util.c [hunk 2]
```diff
@@ -65,7 +66,59 @@ char *js_array_join(JSContext *context, JSValue array, const char *separator) {
...
 
-void js_error_to_string(JSContext *context, JSValue error, char **out, char **out_stack) {
+char *js_error_message_join(const char *name,
+                            size_t name_length,
+                            const char *message,
+                            size_t message_length,
+                            const char *stack,
+                            int include_name,
+                            size_t *out_length) {
+    size_t stack_length = stack != NULL ? strlen(stack) : 0;
```
[the remark this answers is not in the document]
Comment (985892345) [reply quickjs/native/jni/js_value_util.c:76]: Thanks for flagging this. We reviewed the boundary and are intentionally keeping stack values C-string based in this PR. QuickJS-generated stacks do not contain embedded NUL bytes; this requires manually replacing the non-standard stack property. Propagating stack lengths through both string and legacy array-stack extraction would broaden the change beyond the supported error-message fix, so this edge case is out of scope.

File: quickjs/native/jni/mapping/js_value_to_jobject.c [hunk 3]
```diff
@@ -39,86 +42,106 @@ jthrowable js_error_to_java_error(JNIEnv *env, JSContext *context, JSValue error
...
-
     if (java_error_cls != NULL) {
         jmethodID constructor = (*env)->GetMethodID(env, java_error_cls, "<init>",
                                                     "(Ljava/lang/String;)V");
         jthrowable e = try_catch_java_exceptions(env);
         if (e == NULL && constructor != NULL) {
-            jstring java_message = (*env)->NewStringUTF(env, message);
+            jstring java_message = jni_string_from_utf8(env, message, message_length);
             jthrowable java_error = (*env)->NewObject(env, java_error_cls, constructor,
                                                       java_message);
```
[the remark this answers is not in the document]
Comment (985892345) [reply quickjs/native/jni/mapping/js_value_to_jobject.c:107]: Fixed in 9a65939. We preserve the pending OutOfMemoryError rather than clearing it: error construction now returns immediately when JNI string conversion returns NULL, and every caller checks the throwable before calling Throw. This avoids NewObject with a pending exception and Throw(NULL). The full JVM suite and Android Debug/Release builds pass.

**Comment by 985892345** [comment]
Addressed the latest CodeRabbit outside-diff finding in 6e2da01. Object-to-map conversion now checks for a pending JNI exception before Map.put; on conversion failure it releases the current key, remaining property atoms, native property storage, and the partial Java map, then returns without making further JNI calls. The pending OutOfMemoryError is preserved. The full JVM suite and Android Debug/Release builds for all four ABIs pass.

**Comment by dokar3** [comment]
Thanks for the PR! 

LGTM.
